src/Numerics/SoaSphericalTensor.h

```diff
@@ -55,21 +59,58 @@ struct SoaSphericalTensor
   SoaSphericalTensor(const SoaSphericalTensor& rhs) = default;
 
   ///compute Ylm
-  void evaluate_bare(T x, T y, T z, T* Ylm) const;
+  static void evaluate_bare(T x, T y, T z, T* Ylm, const int Lmax_, const T* FacL, const T* FacLM);
 
   ///compute Ylm
   inline void evaluateV(T x, T y, T z, T* Ylm) const
   {
-    evaluate_bare(x, y, z, Ylm);
+    evaluate_bare(x, y, z, Ylm, Lmax, FactorL.data(), FactorLM.data());
     for (int i = 0, nl = cYlm.size(); i < nl; i++)
       Ylm[i] *= NormFactor[i];
   }
 
+  /**
+   * @brief evaluate V for multiple electrons and multiple pbc images
+   * 
+   * @param [in] xyz electron positions [Nelec, Npbc, 3(x,y,z)]
+   * @param [out] Ylm Spherical tensor elements [Nelec, Npbc, Nlm]
+  */
+  inline void batched_evaluateV(OffloadArray3D& xyz, OffloadArray3D& Ylm) const
+  {
+    const size_t nElec = xyz.size(0);
+    const size_t Npbc  = xyz.size(1); // number of PBC images
+    assert(xyz.size(2) == 3);
+
+    assert(Ylm.size(0) == nElec);
+    assert(Ylm.size(1) == Npbc);
+    const size_t Nlm = Ylm.size(2);
+
+    size_t nR = nElec * Npbc; // total number of positions to evaluate
+
+    auto* xyz_devptr     = xyz.device_data();
+    auto* Ylm_devptr     = Ylm.device_data();
+    auto* flm_ptr        = FactorLM.data();
+    auto* fl_ptr         = FactorL.data();
+    auto* NormFactor_ptr = NormFactor.data();
+
+
+    PRAGMA_OFFLOAD("omp target teams distribute parallel for \
+                    map(always, to:flm_ptr[:Nlm], fl_ptr[:Lmax+1], NormFactor_ptr[:Nlm]) \
+                    is_device_ptr(xyz_devptr, Ylm_devptr)")
```

Comment:
Are `FactorLM` and `FactorL` constants? If so, put them in dual space and transfer ahead.